### ISSUE TITLE
Refactor UI classes

### ISSUE THREAD

**Issue by phobos2077** [falltergeist/falltergeist#431]
Issues and solutions:
* Positional properties are not obvious. Need to define solid meaning for `position(), offset(), size()` and the like, validate code and add comments to them. What is position? Is it always screen position? It should be clear.
* New read-only property like `screenPos()` should always return UI position on the screen, exactly where it is drawn (to be used in `render()`).
* Values of those properties are inconsistent. Getters and setters made virtual and some classes override them and change their meaning (like `Animation::offset()`) - it's bad. 
* Some properties have different meaning for values passed to setter and returned from getter. Best example - `position` property. Setter changes internal `_position` member, but getter returns `_position + _offset`. So if you write `img->setPosition(img->position() + Point(10, 20))` you might be surprised by result! It's an indication of flawed design. IMO, what you "get" should be more or less what you "set" (maybe value is capped or validated in setter, but it should have the same meaning - reference point).
* Using base classes is a bad design and we already see some kludges to "hide" base class functionality in subclasses. Need to use interface (like `UI::ElementInterface`). For some classes we can still use common base class (like Image -> ImageButton), but from what I saw it would not be common.

**Comment by phobos2077** [comment]
See bug #421.

**Comment by alexeevdv** [comment]
![default](https://cloud.githubusercontent.com/assets/597839/10493154/53508740-72b8-11e5-8d25-5a9d40abc0cb.png)

Object position in Fallout is tied up to hexagon position. So
```C++
Point object.position = hexagon.position // point A
Point object.ui.position  = object.position + Point(-object.position.width/2, -object.position.height) + object.ui.offset // point B
```
object.ui.offset is used for animation, effects and ui position tuning(offsets in FRM files)

**Comment by phobos2077** [comment]
That's a good illustration, but we still need to get our code in order :)

* Object `position` can be changed, but it will be reset on the next render regardless. Maybe render method should accept position instead? Caller decides where it should render the object, because it usually owns it (hexagon grid own objects, panels own buttons and other controls).
* Animations actually don't use offset which is set by `setOffset()` (they use frame offsets instead). So again, several types of "offsets" - easy to get lost :) 
I think offset() property should be removed altogether from most classes to simplify interfaces.

**Comment by alexeevdv** [comment]
Well... it is pretty simple for me
1) Object(Game::Object) should have no position. It position is defined by hexagon
2) UI has ```position``` and ```offset```. 
3) Animation is UI instance. So it has ```position``` and ```offset``` too. Each frame position is calculated with: ```position + offset + frameOffset```

but i agree. offset() method MUST return value that was set by setOffset() method. There is some mess in our code for now

**Comment by phobos2077** [comment]
I still think we need to remove `offset` from base UI - it's not always needed, and make `position` relative to parent and always reliably settable.
Only other place where offset is used (that I can think of) is dragging inventory items. But current item dragging design is flawed anyway - we need to create new UI for item currently being dragged - it will be moved using `setPosition()` and drawn after all other items.

**Comment by alexeevdv** [comment]
UI position is always relative. In common case it relative to screen. It actual position is calculated in related ```render()``` method

**Comment by phobos2077** [comment]
OK, but now position() property is useless, it gets changed in "render" method (which is very strange, rendering should not mutate objects at all) and it's getter returns `_position + _offset`. That's why I propose to make it just a simple get/set property and either add const Point& argument to Render method or add some "parent" reference and use it's position recursively to calculate exact screen position. (first solution seems better to me, no need for "parent" reference).

**Comment by alexeevdv** [comment]
yep. ```position``` property should stay unchanged during rendering. Also getter must return only ```position``` value without offsets

**Comment by phobos2077** [comment]
Finally we agreed on something. That's progress :+1:

**Comment by alexeevdv** [comment]
But i still believe that there is no reason for ```parent``` property. Each parent must store it children in some kind of vector. In render method parent should append it position to each child. Children position must be relative to parent position.

**Comment by phobos2077** [comment]
I also don't want to add `parent` property, the question is how to append parent position to child position without mutating children? The only solution is to add argument to `render()` method.
